### PR TITLE
feat: Dapo predictor

### PR DESCRIPTION
**Pull request by ml1019-lmx**
# Predictor-Driven Prompt Reordering for DAPO

## Summary

Add predictor-driven prompt reordering to the DAPO training pipeline for the release/v0.7.1 branch. Before each rollout generation, prompts are scored by a lightweight linear predictor (trained via ListMLE loss to predict response length), then reordered with serpentine (snake) DP packing so that high-scoring prompts are distributed evenly across DP ranks.

## Key Files

- `predictor_worker.py` — Predictor actor worker: hidden state extraction, score computation, and ListMLE training
- `predictor_dapo_trainer.py` — Trainer that injects predictor scoring + snake-sort reorder into the training loop
- `predictor_utils.py` — Serpentine sort utility for balanced DP packing

## Experimental Setup

| Parameter | Value |
|-----------|-------|
| Model | Qwen3-30B-A3B-Instruct-2507 |
| DataLoader Seed | 1 |
| Global Batch Size (GBS) | 32 |
| N Samples per Prompt | 8 |
| Max Num Sequences | 16 |
| Generation TP | 4 |
| Sequence Parallel (SP) | 4 (ulysses) |
| Max Model Len | 22528 |
| Prompt Length | ~2k |
| Response Length | ~20k |
| NPU Count | 32 |
| Training Steps | 57 |

### Critic Score

| Metric | Reorder (this PR) | Baseline |
|--------|-------------------|----------|
| Average | 0.6179 | 0.6137 |
| First 10 steps avg | 0.4383 | 0.4391 |
| Last 10 steps avg | 0.6680 | 0.6680 |

**Conclusion:** Critic Score is essentially identical — reordering does not degrade training quality.

### Step Time (s/it)

| Metric | Reorder | Baseline |
|--------|---------|----------|
| Average | 638.98 | 668.40 |
| First 10 steps avg | 616.14 | 621.21 |
| Last 10 steps avg | **616.23** | **711.55** |

**Conclusion:** Reorder version stays stable at ~616 s/it throughout training. Baseline degrades from 621 s/it to 711 s/it — the gap widens from 5.08s to 95.33s.

### Generation Time (s)

| Metric | Reorder | Baseline |
|--------|---------|----------|
| Average | 471.66 | 504.67 |
| First 10 steps avg | 439.39 | 461.45 |
| Last 5 steps avg | **421.14** | **522.81** |
| Trend | -18.25s (decreasing) | +61.37s (increasing) |

**Conclusion:** Reorder version gen time continuously decreases during training; baseline gen time significantly increases. The gap widens from 22s to 101.67s.

### Actor Entropy

| Metric | Reorder | Baseline |
|--------|---------|----------|
| Average | 0.2664 | 0.2626 |
| First 10 steps avg | 0.2571 | 0.2577 |
| Last 5 steps avg | 0.2619 | 0.2611 |
| Trend | +0.0048 (slight increase) | +0.0034 (slight increase) |

**Conclusion:** Both versions have basically the same entropy — no significant difference.

## Key Findings

1. **No quality loss** — Critic Score unchanged (0.6179 vs 0.6137), confirming reordering preserves training quality.
2. **Stable step time** — Reorder stays at ~616 s/it; baseline degrades from 621s to 711s.
3. **Faster generation** — Reorder gen time drops from 439s to 421s; baseline rises from 461s to 523s.
4. **Entropy basically consistent** — Both versions show similar entropy (0.2664 vs 0.2626), no significant difference.
5. **Widening gap** — The gen time advantage grows from 22s to 102s as training progresses.

## How to Enable

```bash
python3 -m recipe.dapo_predictor.main_dapo_predictor_reorder \
...
...
+trainer.predictor_reorder.enable=True \
+trainer.predictor_reorder.epochs=10 \
+trainer.predictor_reorder.batch_size=32 \
+trainer.predictor_reorder.lr=3e-5
```
Co-authored-by @Mind-s

### CODE REVIEW

File: dapo_predictor/README.md [hunk 1]
```diff
@@ -0,0 +1,28 @@
+# DAPO Predictor Reorder (Portable Copy)
+
```
Comment (wucong25) [inline dapo_predictor/README.md:2]: readme说明过于简略，需要加上一些说明文档和使用文档

File: dapo_predictor/predictor_dapo_trainer.py [hunk 2]
```diff
@@ -0,0 +1,693 @@
...
+
+                is_last_step = self.global_steps >= self.total_training_steps
+
+                with marked_timer("step", timing_raw):
+                    with marked_timer("predictor_score", timing_raw, "purple"):
+                        with marked_timer("predictor_hydrate", timing_raw, "purple"):
+                            predictor_input_batch = gen_batch_output.select(deepcopy=True)
+                            predictor_input_batch = self._hydrate_gen_batch_model_inputs(predictor_input_batch)
+                        predictor_order = self._build_predictor_order(predictor_input_batch)
+                        # predictor_scores = self.actor_rollout_wg.compute_predictor_score(predictor_input_batch)
```
Comment (wucong25) [inline dapo_predictor/predictor_dapo_trainer.py]: 无关注释可以删除
